src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Code.java

```diff
@@ -1811,7 +1811,7 @@ State join(State other) {
                     t==tother ? t :
                     types.isSubtype(t, tother) ? tother :
                     types.isSubtype(tother, t) ? t :
-                    error();
+                    commonSuperClass(t, tother);
```

Comment:
Are the `isSubtype()` or `t==tother` checks above still necessary? Like do those fast paths not exist in `types::lub`?

Reply:
Just checked, lub is a varargs call and is slow. However, I think the fast path merge logic `t==tother? ... : commonSuperClass(t, tother)` should be in `commonSuperClass`.

Reply:
The `isSubtype` serve two purposes - they handle `null`/`BOT` type (which, I think, `lub` normally does not handle), and they are cheaper to run.

I've moved them into `commonSuperClass`, as suggested, here:
https://github.com/openjdk/jdk/pull/24617/commits/bb08b490108556553e55b0361f57c99b91a3f89c